crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,288 @@
+                continue;
+            }
+            tracing::debug!("Pulling layer {}", &layer.digest);
+            let mut bytes = Vec::new();
+            self.oci
+                .pull_blob(&reference, &layer.digest, &mut bytes)
+                .await?;
+
+            match layer.media_type.as_str() {
+                WASM_LAYER_MEDIA_TYPE => self.cache.write_wasm(&bytes, &layer.digest).await?,
+                _ => self.cache.write_data(&bytes, &layer.digest).await?,
+            }
+        }
+
+        tracing::info!("Pulled {}@{}", reference, digest);
+
+        Ok(())
+    }
+
+    pub async fn descriptor(&mut self, reference: &str) -> Result<SpinOciDescriptor> {
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+
+        let manifest_path = self
+            .cache
+            .manifest_for_reference(&reference.to_string())
+            .await?;
+        let config_path = self
+            .cache
+            .config_for_reference(&reference.to_string())
+            .await?;
+        if !manifest_path.exists() || !config_path.exists() {
+            self.pull(&reference.to_string()).await?;
+        }
+
+        let config = serde_json::from_slice(&Self::data(&config_path).await?)?;
+        let manifest = serde_json::from_slice(&Self::data(&manifest_path).await?)?;
+
+        Ok(SpinOciDescriptor {
+            reference: reference.to_string(),
+            manifest,
+            config,
+        })
+    }
+
+    /// Create a new wasm layer based on a file.
+    pub async fn wasm_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading wasm module from {:?}", file);
+        Ok(ImageLayer::new(
+            Self::data(file).await.context("cannot read wasm module")?,
+            WASM_LAYER_MEDIA_TYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Create a new data layer based on a file.
+    pub async fn data_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading data file from {:?}", file);
+        Ok(ImageLayer::new(
+            Self::data(&file).await?,
+            DATA_MEDIATYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Construct the registry authentication based on the reference.
+    pub fn auth(reference: &Reference) -> Result<RegistryAuth> {
+        let server = reference
+            .resolve_registry()
+            .strip_suffix("/")
+            .unwrap_or_else(|| reference.resolve_registry());
+
+        match docker_credential::get_credential(server) {
+            Err(CredentialRetrievalError::ConfigNotFound) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::NoCredentialConfigured) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::ConfigReadError) => Ok(RegistryAuth::Anonymous),
+            Err(e) => bail!("Error handling docker configuration file: {:?}", e),
```

Comment:
CredentialRetrievalError implements fmt::Display. We should call that so we get more helpful errors. Right now this is what I see:

Error: Error handling docker configuration file: HelperFailure

If we called .display(), I think we'd get this:

Error: Error handling docker configuration file: Credential helper returned non-zero response code

Reply:
https://github.com/keirlawson/docker_credential/blob/622bef657e2a0bc83abf1ac3f4ab3ed18f9ebf27/src/lib.rs#L24